File: 2019/src/main/kotlin/day09/day9.kt
package day09

import java.io.File
import intcode.IntCodeMachine

fun evaluatorOne(input: String): Long = IntCodeMachine(input).run(1).single()
fun evaluatorTwo(input: String): Long = IntCodeMachine(input).run(2).single()

fun readLinesFromFile(filePath: String): Result<List<String>> =
    runCatching { File(filePath).readLines() }

fun main() {
    readLinesFromFile("src/main/resources/day09.txt")
        .onSuccess {
            val line = it.first()
            println("Part One: ${evaluatorOne(line)}")
            println("Part Two: ${evaluatorTwo(line)}")
        }
        .onFailure {
            println("Error reading file: ${it.message}")
        }
}

File: 2016/src/main/scala/day09/day9.scala
package day09

import scala.util.{Try, Success, Failure, Using}
import scala.io.Source
import scala.util.control.Breaks._

def number(slice: String): (String, Long) = {
    val (digits, rest) = slice.span(_.isDigit)
    return (rest.tail, digits.toLong)
}

def decompress(input: String, partTwo: Boolean): Long = {
    var slice = input
    var length = 0L

    breakable {
        while (true) {
            val start = slice.indexOf('(')
            if (start == -1) break()

            val (temp1, amount) = number(slice.drop(start + 1))
            val (temp2, repeat) = number(temp1)
            val (first, second) = temp2.splitAt(amount.toInt)

            val result = if (partTwo) decompress(first, true) else amount

            slice = second
            length += start + result * repeat
        }
    }

    return length + slice.size
}

def evaluatorOne(input: String): Long = decompress(input, false)
def evaluatorTwo(input: String): Long = decompress(input, true)

def readLinesFromFile(filePath: String): Try[List[String]] =
    Using(Source.fromResource(filePath))(_.getLines().toList)

def hello(): Unit = {
    readLinesFromFile("day09.txt") match {
        case Success(lines) => {
            println(s"Part One: ${evaluatorOne(lines.head)}")
            println(s"Part Two: ${evaluatorTwo(lines.head)}")
        }
        case Failure(exception) => {
            println(s"Error reading file: ${exception.getMessage}")
        }
    }
}
